src/librustdoc/clean/utils.rs

```diff
@@ -562,3 +547,18 @@ crate fn has_doc_flag(attrs: ty::Attributes<'_>, flag: Symbol) -> bool {
 ///
 /// Set by `bootstrap::Builder::doc_rust_lang_org_channel` in order to keep tests passing on beta/stable.
 crate const DOC_RUST_LANG_ORG_CHANNEL: &'static str = env!("DOC_RUST_LANG_ORG_CHANNEL");
+
+/// Render a sequence of macro arms in a format suitable for displaying to the user
+/// as part of an item declaration.
+pub(super) fn render_macro_arms(matchers: &[TokenTree]) -> String {
+    matchers
+        .iter()
+        .map(|matcher| format!("    {} => {{ ... }};\n", render_macro_matcher(matcher)))
+        .collect::<String>()
+}
+
+/// Render a macro matcher in a format suitable for displaying to the user
+/// as part of an item declaration.
+pub(super) fn render_macro_matcher(matcher: &TokenTree) -> String {
+    rustc_ast_pretty::pprust::tt_to_string(matcher)
+}
```

Comment:
Why have this as a separate function? Just use tt_to_string.

Reply:
We may not use `tt_to_string` in the future, or we may have additional logic, so I'd rather err on the side of having an extra function. But I can remove it if you prefer.

Reply:
Please remove it until and unless we add more logic.

Reply:
Actually, `render_macro_matcher` is used in another place, for rendering macros 2.0. I'd really rather keep it so there aren't multiple places to update later on.

Reply:
I don't understand why you think this would ever change. Rustdoc uses rustc in *lots* of places, it doesn't make sense to wrap every single one in our own function.

Reply:
Oh I see, this is the function you mentioned above for removing the whitespace in the matcher. I think we should keep the function if we keep your special-casing, and remove it if not. And I do like that the special-casing makes it prettier. So I guess I'm in favor of keeping it.

Reply:
See also the discussion in https://github.com/rust-lang/compiler-team/issues/403